staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go

```diff
@@ -599,6 +603,22 @@ func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
 		s.HandlerChainWaitGroup.Wait()
 	}()
 
+	if s.ShutdownLongRunningRequestsTimeout != time.Duration(0) {
+		go func() {
+			defer klog.V(1).InfoS("[graceful-termination] shutdown event - draining timed out", "name", drainedCh.Name())
+			defer drainedCh.Signal()
```

Comment:
Line 585 alsi signals `defer drainedCh.Signal()`

since the other goroutine has the `s.HandlerChainWaitGroup.Wait()` , can we really signal here that all request were drained?

Reply:
The idea here was to introduce hard deadline for draining. There are cases when requests can hang "infinitely" and not respect request-timeout. In general I agree that we should fix these issues, like https://github.com/kubernetes/kubernetes/issues/111886 but introducing timeout for draining seems to be useful as well. Maybe I should split it into separate flag?

Reply:
I see, so ShutdownLongRunningRequestsTimeout will define the timeout for DrainedCh ... we have to update the diagram https://github.com/kubernetes/kubernetes/pull/113546/files#diff-68eb390e148644c38174960be4d103c0e377e71eb540aa8c395f7d2278363973R454-R471  in Lines 4054-471 then

Reply:
we have another diagram also in `pkg/server/genericapiserver_graceful_termination_test.go` and we should add a test here to cover this scenario, "hanging request doesn't stop shutdown with ShutdownLongRunningRequestsTimeout set`

Reply:
Done.